File: src/kod/executor.py
"""Executor for step-based rebuild and install workflows.

Runs steps from the planner in order, dispatching by kind, firing lifecycle hooks,
managing per-step timeouts, and enforcing on_error policies.

Spec: docs/superpowers/specs/2026-09-10-kod-planner-hooks-buildcache-design.md §5.
"""

import logging
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

from kod.bootstrap import _convert_to_lua_table
from kod.common import exec, exec_chroot
from kod.planner import Step
from kod.system.boot import get_kernel_file

logger = logging.getLogger(__name__)


class StepError(Exception):
    """Raised when a step fails and on_error='abort'."""

    pass


@dataclass(frozen=True)
class StepResult:
    """Result of executing a single step.

    Immutable record of what happened when a Step ran: whether it succeeded,
    what error (if any) occurred, and metadata for debugging/logging.

    Attributes:
        step: The Step that was executed (see planner.py).
        success: True if the step completed without error.
        error: Error message if step failed (on_error='abort') or warned
            (on_error='warn'). None if success=True.
        stdout: Captured stdout from step execution. None until implemented.
        stderr: Captured stderr from step execution. None until implemented.
        is_warning: True if step failed but on_error='warn', so execution
            continued. False for abort-steps or successful steps.
    """

    step: Step
    success: bool
    error: str | None = None
    stdout: str | None = None
    stderr: str | None = None
    is_warning: bool = False


def execute_steps(
    steps: list[Step],
    mount_point: str,
    use_chroot: bool,
    repos: dict | None = None,
    hooks: dict[str, list[Callable]] | None = None,
) -> list[StepResult]:
    """Run steps via the Lua generic runner (kod/lib/executor.lua).

    Lua handles ordering, on_error policy, hooks, and chroot-wrapped shell
    steps (timeout-guarded, no output capture — lupa disables io.popen).
    Package/service/named-system steps dispatch to a Python callable built
    from repos; the system modules behind them move to Lua in a later phase.

    Raises StepError on abort-step failure or unknown kind.
    """
    import os

    import lupa

    from kod.lua_runtime import get_lua_runtime

    lua = get_lua_runtime()
    base_path = os.path.dirname(os.path.dirname(__file__))
    lua.execute(
        f"package.path = '{base_path}/?.lua;{base_path}/?/init.lua' .. package.path"
    )

    # Force reload of planning modules to ensure we have fresh code
    from kod.lua_runtime import LuaRuntimeManager

    manager = LuaRuntimeManager()
    manager.reload_modules([r"kod\.planning\..*"])

    steps_lua = _convert_to_lua_table(lua, [s.to_dict() for s in steps])

    ctx_data = {
        "mount_point": mount_point,
        "use_chroot": bool(use_chroot),
    }
    if repos is not None:
        ctx_data["repos"] = repos
    ctx_lua = _convert_to_lua_table(lua, ctx_data)

    dispatch_lua = lua.table()
    
    def dispatch_step(step, ctx_lua):
        """Dispatch package/service steps to Python handlers.
        
        Package and service steps have commands that need to be run with
        proper permissions. During install (chroot=True), they run in the chroot.
        During rebuild (chroot=False), they run on host with sudo.
        Returns None (success) or raises on failure.
        """
        try:
            kind = step.kind
            # Command is stored in meta since we don't want it in the step dict
            # (which would make Lua executor treat it as a shell step)
            meta = step.meta if step.meta is not None else {}
            cmd = meta.get("command") if meta else ""
            if not cmd:
                cmd = ""
            
            if kind == "package" or kind == "service":
                if not cmd:
                    return  # No-op step
                
                mp = ctx_lua.mount_point if ctx_lua else "/"
                # Safely access chroot field with default False
                chroot_needed = False
                try:
                    chroot_needed = bool(step.chroot)
                except:
                    chroot_needed = False
                
                if chroot_needed and mp != "/":
                    # During install: chroot into new mount point
                    exec_chroot(cmd, mount_point=mp)
                elif chroot_needed:
                    # chroot=True but mount_point is "/", run as-is (on current system)
                    exec(cmd)
                else:
                    # During rebuild: run on host with sudo
                    exec(f"sudo {cmd}")
            else:
                raise StepError(f"Unknown dispatch step kind: {kind}")
        except StepError:
            raise
        except Exception as e:
            import traceback
            traceback.print_exc()
            raise StepError(f"dispatch_step failed: {e}")
    
    dispatch_lua["step"] = dispatch_step
    
    # Inject Python exec functions for Lua system modules to call back
    # These are used by src/lua/kod/system/boot.lua
    dispatch_lua["_dispatch_python"] = lua.table()
    dispatch_lua["_dispatch_python"]["exec"] = lambda cmd, get_output=False: exec(cmd, get_output=get_output)
    dispatch_lua["_dispatch_python"]["exec_chroot"] = lambda cmd, **kw: exec_chroot(cmd, **kw)
    dispatch_lua["_dispatch_python"]["get_kernel_file"] = lambda mp, pkg="linux": get_kernel_file(mp, pkg)

    hooks_data = hooks or {}
    hooks_lua = _convert_to_lua_table(lua, hooks_data)

    module_result = lua.require("kod.planning.executor")
    # lupa returns a tuple (module, ...) when requiring Lua modules
    if isinstance(module_result, tuple):
        module = module_result[0]
    else:
        module = module_result
    try:
        results_lua = module.run(steps_lua, ctx_lua, dispatch_lua, hooks_lua)
    except lupa.LuaError as e:
        raise StepError(str(e))

    results = []
    for i in range(1, len(results_lua) + 1):
        r = results_lua[i]
        results.append(
            StepResult(
                steps[i - 1],
                bool(r.success),
                error=r.error,
                is_warning=bool(r.is_warning),
            )
        )
    return results

File: src/lua/kod/planning/executor.lua
-- Generic step runner for KodOS plans (language-agnostic orchestrator).
-- Runs steps in order, fires pre/post hooks, applies on_error policies.
--
-- Step kinds and execution paths:
--   system/disk: Run as shell commands with `timeout` guard and optional chroot wrap.
--                Uses os.execute for direct shell invocation (no output capture; lupa disables io.popen).
--   lua-system:  Dispatch to Lua system modules (e.g., kod.system.boot) via step.meta.operation.
--                Lua modules call Exec.exec_chroot() wrapper, which routes to Python exec_chroot
--                for kernel operations (distro-specific logic remains in Python for now).
--   package/service: Dispatch to host-side callable dispatch.step(step, ctx).
--                    May return "shell" to fall back to os.execute, or handle the step directly.
--
-- Architecture note:
--   - Executor.run_shell uses os.execute: generic shell runner (no output capture, no dependencies).
--   - Lua system modules (boot.lua) use Exec.exec_chroot: proper error handling and output capture
--     for critical boot operations that need Python's distro integration layer.

local Executor = {}

-- Quote a string for embedding in sh -c '...'
-- Uses the standard shell technique: '...' is literal, so to include a quote,
-- close quote, add escaped quote, reopen: 'foo'"'"'bar' = foo'bar
local function shq(s)
    return "'" .. s:gsub("'", "'\\''") .. "'"
end

-- Run a shell step (system or disk). Returns {success=...} or {success=false, error=...}.
-- Uses os.execute for direct shell invocation (no output capture due to lupa restrictions).
-- For operations requiring Python integration (kernel setup, etc.), use lua-system kind instead,
-- which dispatches to Lua system modules that call Exec.exec_chroot() wrapper.
function Executor.run_shell(step, mount_point)
    local parts = { step.program or step.command or "" }
    for i = 1, #(step.args or {}) do
        table.insert(parts, step.args[i])
    end
    local cmd = table.concat(parts, " ")
    if step.chroot then
        cmd = "chroot " .. shq(mount_point or "/") .. " sh -c " .. shq(cmd)
    end
    -- coreutils `timeout` guards each step; exit 124 on expiry.
    local full = "timeout " .. (step.timeout_s or 300) .. " " .. cmd
    local ok, term, status = os.execute(full)
    if ok then
        return { success = true }
    end
    return { success = false, error = "command failed (status " .. tostring(status) .. "): " .. cmd }
end

-- steps:   array of step tables {kind,name,program,args,chroot,timeout_s,on_error,meta}
-- ctx:     table with mount_point, use_chroot, repos
-- dispatch: table with dispatch.step(step, ctx) for package/service steps (optional)
-- hooks:   table mapping "pre:<kind>"/"post:<kind>" to arrays of callables
--
-- Returns: array of result tables {success, error, is_warning}
function Executor.run(steps, ctx, dispatch, hooks)
    hooks = hooks or {}
    local results = {}

    for i = 1, #steps do
         local step = steps[i]

         -- pre hooks: error aborts the whole run
        for _, hook in ipairs(hooks["pre:" .. step.kind] or {}) do
            local ok, err = pcall(hook, step, ctx)
            if not ok then
                error("pre:" .. step.kind .. " hook failed for step '" .. step.name .. "': " .. tostring(err))
            end
        end

        local result
        -- Package/service steps dispatch to host callbacks (which handle sudo/root)
        -- even if they have commands; other steps with commands are shell steps
        if step.kind == "package" or step.kind == "service" then
            -- Dispatch to host: the host decides whether to run via shell, sudo, or skip
            if dispatch and dispatch.step then
                local ok, r = pcall(dispatch.step, step, ctx)
                if not ok then
                    result = { success = false, error = tostring(r) }
                elseif r == "shell" then
                    result = Executor.run_shell(step, ctx.mount_point)
                else
                    result = { success = true }
                end
            else
                result = { success = true }
            end
        elseif (step.program or step.command) and (step.program or step.command) ~= "" then
            -- Any other step carrying a command is a shell step (system or disk).
            result = Executor.run_shell(step, ctx.mount_point)
        elseif step.kind == "lua-system" then
            -- Lua system module: call the appropriate function from the boot module
            local ok, err = pcall(function()
                local boot = require('kod.system.boot')
                local op = step.meta and step.meta.operation
                local kernel = step.meta and step.meta.kernel
                local generation = step.meta and step.meta.generation
                
                if op == "update_kernel" then
                    boot.update_kernel(kernel, ctx.mount_point)
                elseif op == "update_initramfs" then
                    boot.update_initramfs(kernel, ctx.mount_point)
                elseif op == "create_boot_entry" then
                    boot.create_boot_entry(generation, kernel, ctx.mount_point)
                else
                    error("Unknown lua-system operation: " .. tostring(op))
                end
            end)
            
            if not ok then
                result = { success = false, error = "lua-system step failed: " .. tostring(err) }
            else
                result = { success = true }
            end
        else
            result = { success = false, error = "Unknown step kind: " .. tostring(step.kind) }
        end

        if not result.success then
            if step.on_error ~= "warn" then
                error(result.error or ("step '" .. step.name .. "' failed"))
            end
            result.is_warning = true
        end
        table.insert(results, result)

        -- post hooks: errors are swallowed (parity with Python executor)
        for _, hook in ipairs(hooks["post:" .. step.kind] or {}) do
            pcall(hook, step, ctx)
        end
    end

    return results
end

return Executor
